### DISCUSSION TITLE
[Feature request] fluent static members for records

### DISCUSSION THREAD

**Discussion by Zaid-Ajaj** [MoiraeSoftware/Myriad#62] [Ideas]
Given an input record type:
```fs
namespace Shared.Requests 

[<Generator.FluentStaticMembers>]
type UpdateProduct = {
  productId: int
  name: string option
  categoryId : int option
}
```
I would love to have an attributes that generates the following for me:
```fs
namespace Shared.Requests

[<AutoOpen>]
module UpdateProductExtensions = 
    type UpdateRecord with 
        // create a minimal instance by providing the required fields
        static member Create(productId: int) = 
          { productId = productId; name = None; categoryId: None }  
        
       member product.WithName(name: string) = { product with name = Some name }
       member product.WithName(name: string option) = { product with name = name } 
       member product.WithCategoryId(categoryId: int) = { product with categoryId = Some categoryId }
       member product.WithCategoryId(categoryId: int option) = { product with categoryId = categoryId } 
```
Then one would be able to write
```fs
UpdateProduct
  .Create(productId=123)
  .WithName("Updated name")
  .WithCategoryId(None)
```
> An option that makes the `.With` prefix optional would also be nice: `[<Generator.FluentStaticMembers "With">]`

I think this would be really useful for the [Snowflaqe](https://github.com/Zaid-Ajaj/Snowflaqe) project. I might give this a try myself on one of the streams when time permits :smile:

**Comment by 7sharp9** [comment]
Could you just use the `Generate.Lenses` plugin?

**Comment by 7sharp9** [comment]
You can also use the `create` function that the fields plugin creates, so you could do:
`UpdateProduct.create 123 "Updated Name" None`

**Comment by Zaid-Ajaj** [comment]
I need overloaded static members as extension methods, not module functions because in my use case, there are a lot of optional fields that I don't care about at first.

**Comment by 7sharp9** [comment]
Makes sense.  I often use statics for better option handling.  Maybe this could expand to other areas too.

**Comment by robkuz** [comment]
@Zaid-Ajaj what about this?

    type Foo =
        {
            a: int
            b: Option<int>
            c: List<string>
        }
        member this.Update(?a: int, ?b: int, ?c: List<string>) =
            let res =
                match a with
                | Some v ->
                    {
                        this with a = v 
                    }
                | None -> this
            let res =
                match b with
                | Some v ->
                    {
                        res with b = v |> Some
                    }
                | None -> this
            let res =
                match c with
                | Some v ->
                    {
                        res with c = v 
                    }
                | None -> this
            res
            
    let x =
        {
            a = 1
            b = None
            c = []
        }
            
    let y = x.Update(a = 2, c = ["foo";"bar"])
